cgroups for mpm-itk for Apache 2.2.17-01 v0.5

Usage: AssignUserID USER GROUP [/dev/cgroup/CGROUP/tasks]
Too simple: user or group must be changed to change cgroup.

Signed-off-by: Denis Kaganovich <mahatma@eu.by>

PS Rollback: 0.6 was insecure

diff -pruN a/server/mpm/experimental/itk/itk.c b/server/mpm/experimental/itk/itk.c
--- a/server/mpm/experimental/itk/itk.c	2011-02-28 22:40:33.000000000 +0200
+++ b/server/mpm/itk/itk.c	2011-03-12 22:43:17.000000000 +0200
@@ -164,6 +164,7 @@ typedef struct
     gid_t gid;
     char *username;
     int nice_value;
+    char *pidfile;
 } itk_per_dir_conf;
 
 typedef struct
@@ -1421,6 +1422,7 @@ static int itk_post_perdir_config(reques
     gid_t wanted_gid;
     const char *wanted_username;
     int err = 0;
+    FILE *pidf;
     
     itk_server_conf *sconf =
         (itk_server_conf *) ap_get_module_config(r->server->module_config, &mpm_itk_module);
@@ -1465,6 +1467,21 @@ static int itk_post_perdir_config(reques
     }
 
     if (!err && wanted_uid != -1 && wanted_gid != -1 && (getuid() != wanted_uid || getgid() != wanted_gid)) {
+        if (dconf->pidfile) {
+            pidf = fopen(dconf->pidfile, "a");
+            if (!pidf) {
+                 _DBG("fopen(%s): %s", dconf->pidfile, strerror(errno));
+                 err = 1;
+            } else {
+                 if (fprintf(pidf, "%d\n", getpid()) < 0) {
+                    _DBG("fprintf(%s): %s", dconf->pidfile, strerror(errno));
+                    err = 1;
+                 }
+                 fclose(pidf);
+            }
+            if (err)
+                 goto err;
+        }
         if (setgid(wanted_gid)) {
             _DBG("setgid(%d): %s", wanted_gid, strerror(errno));
             err = 1;
@@ -1483,6 +1500,7 @@ static int itk_post_perdir_config(reques
      * at anytime, so we excercise our right. :-)
      */
     if (err) {
+err:
         ap_log_error(APLOG_MARK, APLOG_WARNING, 0, NULL, \
             "Couldn't set uid/gid/priority, closing connection.");
         ap_lingering_close(r->connection);
@@ -1623,12 +1641,13 @@ static const char *set_server_limit (cmd
     return NULL;
 }
 
-static const char *assign_user_id (cmd_parms *cmd, void *ptr, const char *user_name, const char *group_name)
+static const char *assign_user_id (cmd_parms *cmd, void *ptr, const char *user_name, const char *group_name, const char *pid_file)
 {
     itk_per_dir_conf *dconf = (itk_per_dir_conf *) ptr;
     dconf->username = apr_pstrdup(cmd->pool, user_name);
     dconf->uid = ap_uname2id(user_name);
     dconf->gid = ap_gname2id(group_name);
+    dconf->pidfile = pid_file ? apr_pstrdup(cmd->pool, pid_file) : NULL;
     return NULL;
 }
 
@@ -1676,1 +1695,1 @@ AP_INIT_TAKE1("MaxClients", set_max_clie
-AP_INIT_TAKE2("AssignUserID", assign_user_id, NULL, RSRC_CONF|ACCESS_CONF,
+AP_INIT_TAKE23("AssignUserID", assign_user_id, NULL, RSRC_CONF|ACCESS_CONF,
@@ -1691,6 +1710,7 @@ static void *itk_create_dir_config(apr_p
         apr_pcalloc(p, sizeof(itk_per_dir_conf));
     c->uid = c->gid = -1;
     c->nice_value = UNSET_NICE_VALUE;
+    c->pidfile = NULL;
     return c;
 }
 
@@ -1706,10 +1726,12 @@ static void *itk_merge_dir_config(apr_po
       c->username = child->username;
       c->uid = child->uid;
       c->gid = child->gid;
+      c->pidfile = child->pidfile;
     } else {
       c->username = parent->username;
       c->uid = parent->uid;
       c->gid = parent->gid;
+      c->pidfile = parent->pidfile;
     }
     if (child->nice_value != UNSET_NICE_VALUE) {
       c->nice_value = child->nice_value;
